### ISSUE TITLE
Consider switching to @threads/json-patch-ot for OT

### ISSUE THREAD

**Issue by jasoniangreen** [Palindrom/Palindrom#274]
I have written a new lib based off your json-patch-ot which has the following advantages.

- Completely backward compatible with your lib
- Fully written and typed in TS
- Covers many more cases than existing lib
- We can combine efforts in improving one code base (there are still some cases which could be improved such as transforming changes against move operations, see issues).
- I am currently giving talks about this lib which can give more attention to your project if we unify the libs.

I'm happy to do the work and make a PR, but I wanted to know first if there was interest to do so.

Here it is: https://github.com/ThreadsStyling/json-patch-ot

**Comment by tomalec** [comment]
Hi, thanks for letting us know :)

I'd be more than happy to see our collaboration.

Just to give you a brief background that would help to comment your points:
I implemented JSON-Patch-OT back in 2015, to mitigate obvious limitations of JSON-Patch communication. Since then, I made few talks on it (like https://2015.webcampzg.org/talks/view/real-time-asynchronous-collaboration-on-json/). Also, for all those years it was working surprisingly stable for our use-cases in Palindrom and Starcounter. I even considered writing a formal spec, like RFC for Versioned JSON-Patch & co.

Now to answer points from your README:
> Palindrom obviously needed only a subset of the functionality of JSON Patch as large areas were not implemented. Transforming operations against an `add` to an array, or a `move` into an array are just some of the cases that were not covered at all.

Sure, I skipped it due to lack of time and YAGNI, but would be really happy to have it here. `JSON-Patch-OT` used to be in stand-alone repo, and was moved to Palindrom just to help me and my colleagues focus, but from the very beginning we planned to make it an independent library, not something just for Palindrom.

> Merging back changes into a lib for a framework that didn't need those changes was always going to be awkward. We needed to move fast.

We would definitely review and merge new features, as I mentioned we want to make it library of general use. Plus, Palindrom would eventually need other operations as well.
The limitation in set of supported operations came from the only JSON-Patch server we had at the time - Starcounter. Recently we started implementing Palindorm NodeJS server, which would probably use more operations and scenarios.

Naturally, your business decisions, and optimizations of you workflow are totally up to you, and should meet your needs first.

>    We wanted to write the whole thing in TypeScript to give us a safer more stable core lib.

Valid point. Personally, I prefer to write the code which is executed in browser - JS, for debugging and simpler workflow. However, `fast-json-patch` which we also use and maintain is written in TS, so thas's something we could definitely consider. For sure we can start with type definitions.

>    There hadn't been much activity on the lib in a while when we were looking at it.

Because it was working extremely stable for our use-cases, and nobody from outside our team posted any issue about it 😄  Probably if you would, you would see the response quite fast.

Where are you going to give you talks. I'd like to hear it and it would be cool to meet and chat in person.


Speaking of PRs, it think PR to `JSON-Patch-OT` would be easier to digest. Also, I'm working on porting Palindrom to ES Modules, so making a PR here could result in a lot of conflicts.

**Comment by jasoniangreen** [comment]
Yes, I totally understand regarding YAGNI and not needing to make very frequent changes to the code base. I actually only made changes to ours recently when preparing an example for a talk that happened to hit an edge case.

The talk I gave recently should be up here in a few days: https://www.youtube.com/channel/UC-qjgj2TnRoI1ZmFFaN3FeQ

I'll take a look at the `JSON-Patch-OT` repo then.

**Comment by jasoniangreen** [comment]
By the way @tomalec here is the talk I gave at LNUG not long ago.

https://www.youtube.com/watch?v=sLwjXjjZ9Hs&feature=youtu.be

Most of the content won't be anything new to you of course :) And apologies if I'm wrong about anything I talk about, particularly when it comes to reasons we didn't use any existing technologies/solutions. I am ready to put my hand up and say we didn't do a lot of investigation into Palindrome, but that can also be related to there not being enough examples and docs that were very visible for palindrome around the time we were deciding.

It can be difficult to understand what is abandoned and what is just working for the small group of people using and developing a library 😄 

Happy to talk about it all sometime.

**Comment by tomalec** [comment]
@jasoniangreen thanks for sharing your video, and sorry once again for the late response. 
I had some personal and employment issues to manage.

Getting back to your implementation. Did I get it correctly, that you are using a single version number to resolve conflicts? How come that works? Do you assume, or guarantee otherwise that changes are received in the same order they were sent?

------

Example scenario:
```
// initial state, with which Alice and Bob starts
tree = {...}
// Alice given the initial state sends the first change to Bob
a1 = [
	1, // version number/id
    // in our solution we use also JSON Patch for that
    // {op: 'test', path: '/_Version`, value: 1}
	// actial change
	{op: ...}
]
// the state of Alice's tree is: tree + a1, version: 1
// Alice optimistically continues applying changes
a2 = [
	2,
	{op: ...}
]
// the state of Alice's tree is: tree + a1 + a2, version: 2

// Bob given the initial tree sends his first change to Alice, before he received anything from Alice
b1 = [
	1,
	{op: ...}
]
// the state of tree is: tree + b1, version: 1

// now a1 comes to Bob
// the state of tree is: tree + b1 + a1, version: 1

```
Would the version of second change issued by Bob be 2, regardless of the change number from Alice?

---- 
Then what if:

We start with a tree `T`.
Alice makes `m` changes and sends them to Bob without waiting for his response.

| change | state of Alice's tree |
| ---   | ---       |
| a_1    | `T + a_1`    |
| ...   | ...    |
| a_m    | `T + a_1 + ... + a_m` |

Bob makes a change to the initial tree

| change | state of Bob's tree |
| ---   | ---       |
| b1    | `T + b1`    |

Bob receives the first `n` changes from Alice and transforms them against his own. (`n<m`)

| change | state of Bob's tree |
| ---   | ---       |
| b_1    | `T + b_1`    |
| a_1    | `T + b_1 + a_1*b_1`   |
| ...    | ...   |
| a_n    | `T + b_1 + a_1*b_1 + ... + a_n*b_1`   |

Bob creates second change based on the state he knows so far (`T, b_1, a_1, ..., a_n`)


| change | state of Bob's tree |
| ---   | ---       |
| b_1    | `T + b_1`    |
| a_1    | `T + b_1 + a_1*b_1`   |
| ...    | ...   |
| a_n    | `T + b_1 + a_1*b_1 + ... + a_n*b_1`   |
| b_2    | `T + b_1 + a_1*b_1 + ... + a_n*b_1 + b_2`   |

Alice receives first change from Bob (`b_1`) and transforms it against all her changes

| change | state of Alice's tree |
| ---   | ---       |
| a_1    | `T + a_1`    |
| ...   | ...    |
| a_m    | `T + a_1 + ... + a_m` |
| b_1    | `T + a_1 + ... + a_m + b_1*a_1*...*a_m` |

Alice receives second change from Bob (`b_2`) which was taking `a_1` to `a_n` into account.

How come Alice knows that she should transform `b_2` against `a_(n+1), ..., a_m` and not `a_1, ..., a_n`?

-----

I guess you maight not get hit by this issue as you are not implementing direct collaboration on JSON, like between Alice and Bob. 
You rather have it orchestrated by server: Alice - server - Bob. Then probably you handle change requests synchronously, require request-response flow (unlike in Web Socket push), which guarantees consecutive version number resolution.

But then, how do you guarantee eventual consistency between Alice and server, and Bob and server? :)
